### PR TITLE
CompartmentCounter for tracking number of people in each compartment

### PR DESCRIPTION
**Pull request by NicholasFan235**
Added CompartmentCounter which keeps track of number of people in each compartment

### CODE REVIEW

File: pyEpiabm/pyEpiabm/__init__.py [hunk 1]
```diff
@@ -16,6 +16,7 @@
 from .microcell import Microcell  # noqa
 from .household import Household  # noqa
 from .parameters import Parameters  # noqa
+from .compartment_counter import CompartmentCounter  # noqa
```
Comment (KCGallagher) [inline pyEpiabm/pyEpiabm/__init__.py:19]: @NicholasFan235 can I run this change past you, had second thoughts that we might not want `CompartmentCounter` easily accessible if it's not meant to be called directly? Not really sure what I am talking about here - is this okay?

Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/__init__.py:19]: Okay so will leave this but rename to `_CompartmentCounter`

File: pyEpiabm/pyEpiabm/_compartment_counter.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        :type n_people: int
+        """
+        self._compartments[InfectionStatus.Susceptible] = n_people
+
+    def report(self, old_status: InfectionStatus,
+               new_status: InfectionStatus) -> None:
+        """Report Person has changed state.
+        Update internal compartments state.
+
+        :param old_status: Person's previous infection state.
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/_compartment_counter.py:40]:
```suggestion
        :param old_status: Person's previous infection state
```

File: pyEpiabm/pyEpiabm/_compartment_counter.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        :type old_status: :class:`InfectionStatus`
+        :param new_status: Person's new infection state.
+        :type new_status: :class:`InfectionStatus`
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/_compartment_counter.py:43]:
```suggestion
        :type old_status: InfectionStatus
        :param new_status: Person's new infection state
        :type new_status: InfectionStatus
```

File: pyEpiabm/pyEpiabm/_compartment_counter.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        self._compartments[old_status] -= 1
+        self._compartments[new_status] += 1
+
+    def retrieve(self) -> typing.Dict[InfectionStatus, int]:
+        """Get Compartment Counts
+        Returns dictionary of compartment counts
+
+        :return: compartments
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/_compartment_counter.py:52]:
```suggestion
        """Get Compartment Counts.
        Returns dictionary of compartment counts.

        :return: Compartments
```

File: pyEpiabm/pyEpiabm/cell.py [hunk 5]
```diff
@@ -42,3 +45,24 @@ def enqueue_person(self, person: Person):
         :type person: Person
         """
         self.person_queue.put(person)
+
+    def _setup(self) -> None:
+        """Setup method. Should be called once Population has been setup.
+        Called by population (DOESN'T NEED TO BE CALLED MANUALLY)
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/cell.py:51]:
```suggestion
        Called by population (DOESN'T NEED TO BE CALLED MANUALLY).
```

File: pyEpiabm/pyEpiabm/cell.py [hunk 6]
```diff
@@ -42,3 +45,24 @@ def enqueue_person(self, person: Person):
...
+    def notify_person_status_change(
+            self,
+            old_status: InfectionStatus,
+            new_status: InfectionStatus) -> None:
+        """Notify Cell that a person's status has changed.
+
+        :param old_status: Person's old infection status.
+        :type old_status: :class:`InfectionStatus`
+        :param new_status: Person's new infection status.
+        :type new_status: :class:`InfectionStatus`
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/cell.py:66]:
```suggestion
        :param old_status: Person's old infection status
        :type old_status: InfectionStatus
        :param new_status: Person's new infection status
        :type new_status: InfectionStatus
```

File: pyEpiabm/pyEpiabm/microcell.py [hunk 7]
```diff
@@ -38,3 +42,23 @@ def add_people(self, n):
             p = Person(self)
             self.cell.persons.append(p)
             self.persons.append(p)
+
+    def _setup(self) -> None:
+        """Setup method. Should be called once Population has been setup.
+        Called by population (DOESN'T NEED TO BE CALLED MANUALLY)
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/microcell.py:48]:
```suggestion
        Called by population (DOESN'T NEED TO BE CALLED MANUALLY).
```

File: pyEpiabm/pyEpiabm/microcell.py [hunk 8]
```diff
@@ -38,3 +42,23 @@ def add_people(self, n):
...
+    def notify_person_status_change(
+            self,
+            old_status: InfectionStatus,
+            new_status: InfectionStatus) -> None:
+        """Notify Microcell that a person's status has changed.
+
+        :param old_status: Person's old infection status.
+        :type old_status: :class:`InfectionStatus`
+        :param new_status: Person's new infection status.
+        :type new_status: :class:`InfectionStatus`
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/microcell.py:61]:
```suggestion
        :param old_status: Person's old infection status
        :type old_status: InfectionStatus
        :param new_status: Person's new infection status
        :type new_status: InfectionStatus
```

File: pyEpiabm/pyEpiabm/person.py [hunk 9]
```diff
@@ -73,3 +73,14 @@ def __repr__(self):
         :rtype: str
         """
         return f"Person, Age = {self.age}"
+
+    def update_status(self,
+                      new_status: InfectionStatus) -> None:
+        """Update Person's Infection Status.
+
+        :param new_status: Person's new status.
+        :type new_status: :class:`InfectionStatus`
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/person.py:82]:
```suggestion
        :param new_status: Person's new status
        :type new_status: InfectionStatus
```

File: pyEpiabm/pyEpiabm/person.py [hunk 10]
```diff
@@ -73,3 +73,14 @@ def __repr__(self):
         :rtype: str
         """
         return f"Person, Age = {self.age}"
```
Comment (I-Bouros) [inline pyEpiabm/pyEpiabm/person.py:75]:
```suggestion
        return f"Person, Age = {self.age}."
```

**Review by I-Bouros — CHANGES_REQUESTED** [review]
Minor comments on styling and one of the docstrings.

File: pyEpiabm/pyEpiabm/_compartment_counter.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    def identifier(self):
+        """Get identifier
+        """
+        return self._identifier
+
+    def initialize(self, n_people) -> None:
+        """Initialize Compartments with n_people susceptible and 0 in all the rest.
+
+        :param n_people: Number of people CompartmentCounter
+                         is keeping track of.
```
[the remark this answers is not in the document]
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/_compartment_counter.py:30]: I don't like the dangling preposition here either, so have removed the full stop in a separate change remotely

File: pyEpiabm/pyEpiabm/_compartment_counter.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        self._compartments = {status: 0 for status in InfectionStatus}
+
+    @property
+    def identifier(self):
+        """Get identifier
+        """
+        return self._identifier
+
+    def initialize(self, n_people) -> None:
+        """Initialize Compartments with n_people susceptible and 0 in all the rest.
```
[the remark this answers is not in the document]
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/_compartment_counter.py:27]: Have rephrased remotely

File: pyEpiabm/pyEpiabm/cell.py (hunk 5 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/cell.py:51]: Might remove caps here too as it it won't look good on the online docs

File: pyEpiabm/pyEpiabm/person.py (hunk 10 above)
Comment (KCGallagher) [reply pyEpiabm/pyEpiabm/person.py:75]: Also updated tests to reflect this

File: pyEpiabm/pyEpiabm/cell.py [hunk 13]
```diff
@@ -16,6 +18,7 @@ def __init__(self):
         self.microcells = []
         self.persons = []
         self.person_queue = Queue()
+        self.compartment_counter = _CompartmentCounter(f"Cell {hash(self)}")
```
Comment (rcw5890) [inline pyEpiabm/pyEpiabm/cell.py:21]:
```suggestion
        self.compartment_counter = _CompartmentCounter(f"Cell {id(self)}")
```

**Review by rcw5890 — CHANGES_REQUESTED** [review]

**Review by rcw5890 — APPROVED** [review]

**Pull request merged**